presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java

```diff
@@ -173,17 +184,48 @@ public Node selectCurrentNode()
                 assignmentStats.addAssignedSplit(chosenNode);
                 topologicCounters[chosenDepth]++;
             }
+            else {
+                splitWaitingForAnyNode = true;
+            }
         }
         for (int i = 0; i < topologicCounters.length; i++) {
             if (topologicCounters[i] > 0) {
                 topologicalSplitCounters.get(i).update(topologicCounters[i]);
             }
         }
-        return assignment;
+
+        CompletableFuture<?> blocked;
+        if (existingTasks.isEmpty()) {
+            blocked = completedFuture(null);
+        }
+        else {
+            int maxPendingForWildcardNetworkAffinity = calculateMaxPendingSplits(0, networkLocationSegmentNames.size());
+            if (splitWaitingForAnyNode) {
+                List<CompletableFuture<?>> stateChangeFutures = existingTasks.stream()
+                        .map(remoteTask -> remoteTask.whenSplitQueueHasSpace(calculateLowWatermark(maxPendingForWildcardNetworkAffinity)))
+                        .collect(toImmutableList());
+                blocked = firstCompletedFuture(stateChangeFutures, true);
+            }
+            else {
+                blocked = toWhenHasSplitQueueSpaceFuture(blockedExactNodes, existingTasks, calculateLowWatermark(maxPendingForWildcardNetworkAffinity));
+            }
+        }
+        return new SplitPlacementResult(blocked, assignment);
+    }
+
+    private int calculateMaxPendingSplits(int currentDepth, int totalDepth)
```

Comment:
It's computing how much of the queue is available to a split with the given network distance. Perhaps renaming `currentDepth` to `splitAffinity` would help?

A split with zero affinity `f(0, 3)` can use only half the queue, whereas one that is local to the worker `f(3, 3)` is allowed to fill the entire queue